src/Shared/Buffers.MemoryPool/SlabMemoryPool.cs

```diff
@@ -112,10 +112,14 @@ private MemoryPoolBlock AllocateSlab()
             _slabs.Push(slab);
 
             // Get the address for alignment
-            var gcHandle = GCHandle.Alloc(slab.PinnedArray, GCHandleType.Pinned);
-            var basePtr = gcHandle.AddrOfPinnedObject();
-            // Release the handle; the array is pinned we only need the GCHandle for the address.
-            gcHandle.Free();
+            IntPtr basePtr;
+            unsafe
+            {
+                fixed(byte* ptr = slab.PinnedArray)
+                {
+                    basePtr = (IntPtr)ptr;
```

Comment:
Perhaps `Marshal.UnsafeAddrOfPinnedArrayElement` could be simpler here?

Reply:
TIL :)

Done.

The `alignment` parameter of the pinned allocated doesn't look to be exposed? Otherwise could skip most of the following code also (which 4096 aligns the blocks)

Reply:
That would be amazing

Reply:
Hmm. Since pinned objects do not move, alignment would be fairly simple to arrange internally.  (compacting would be tricky but that is not an issue)

I could just get objSize + 4096 bytes from the heap and then format that as 2 or 3 objects, with one being the actual result with aligned payload, which I would return, and others immediately becoming garbage, up for reuse.

Reply:
That would be an API addition though with all the process attached, as usual.

Reply:
Can you drive that addition @VSadov ?

Reply:
I will try. Alignment was already a part of the original "cadillac" proposal for `AllocateArray`. Having actual use case would make it easier to argue for the usefullness.

The challenge could be to scope this to just pinned allocations. More general support for aligned allocations is possible, but that is completely different cost.